### ISSUE TITLE
Different results between detect.py and detection by onnx

### ISSUE THREAD

**Issue by liquored** [ultralytics/yolov5#11508]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I export an onnx model from my pth. However it got different results between onnx and pth detection. And this is a part of my onnx detection code.

` 
        dataset = LoadImages(img_path, img_size=img_size) 

        img = torch.zeros((1, 3, img_size, img_size), device=device)  # init img

        for img_path, img, img_ori, _ in dataset:

            img = torch.from_numpy(img).to(device)

            img /= 255.0  # 0 - 255 to 0.0 - 1.0

        if img.ndimension() == 3:

            img = img.unsqueeze(0)

        pred = self.onnx_session.run(None, img)[0]`

Looking forward to your reply!
Yours, Bob Zhou!

**Comment by glenn-jocher** [comment]
@liquored hi Bob Zhou,

The issue you are facing could be due to the differences in the implementation of the detection in the PTH and the ONNX model. The ONNX model could use a different implementation of the layers compared to the PTH model. You could also check if there are any differences between the input pre-processing and output post-processing of the two models. It would be useful if you could provide more information such as the versions of PyTorch and ONNX that you are using.

Best regards, 
Glenn Jocher

**Comment by liquored** [comment]
@glenn-jocher hi Glenn Jocher,
Thanks for your reply! I have dealt with the issue, by change the version of pytorch when you try to export the onnx model. I used torch 1.9.1+cu11.1 before, and when I changed it to torch 1.7.1, it works well. I didn't know the reasons of this, but it seems to work! 
Not only that, but it also should attract attention that I should scale the picture before padding, which also troubles me a bit!
Thanks for your early reply, wish this issue could help others!
Best regards,
Bob Zhou

**Comment by glenn-jocher** [comment]
@liquored hi Bob Zhou,

I'm glad to hear that you were able to resolve your issue by changing the version of PyTorch that you were using. The differences in the PyTorch versions could indeed cause differences in the exported ONNX model. It's always a good idea to verify the compatibility between the PyTorch and ONNX versions before export.

Additionally, it's great that you found a solution for scaling before padding, as that can definitely impact the model performance.

Thanks for sharing your experience with this issue, it will definitely be helpful for others facing similar problems.

Best regards,
Glenn Jocher
